### ISSUE TITLE
Check third party libraries

### ISSUE THREAD

**Issue by MarcusRainbow** [MarcusRainbow/QuantMath#9]
We rely on rand for random numbers. Is this as good as Mersenne Twister? (If not, it would be better to implement Mersenne Twister.)

We rely on statrs for Normal distribution functions. How accurate are these? It really matters for far out of the money options how good the cumulative normal function is in the wings.

**Comment by petr-tik** [comment]
Just found this library (amazing work) and I was wondering if you still have any time and interest in accepting PRs? 

If so, there is [a crate implementing the mersenne twister PRNG ](https://docs.rs/mersenne_twister/1.1.1/mersenne_twister/), which I can add behind a feature flag

**Comment by MarcusRainbow** [comment]
Hi,
I've not done anything to this library for a while. However, if you want to
make any contributions, they would be more than welcome. I agree it would
be very good to use Mersenne Twister for the pseudo-random numbers. (Sobol
sequences would also be very useful.)

I still think there is a real future in using Rust for financial maths. I'm
just waiting for the rest of the world to cotton on to this idea...
Thanks,
Marcus

**Comment by petr-tik** [comment]
> I still think there is a real future in using Rust for financial maths. I'm
just waiting for the rest of the world to cotton on to this idea

Hear hear! I guess the hardest problem to solve is acknowledging the moat of C++ and integrating rust with existing codebases as seamlessly as possible. Have you had any success or problems with that? 

> I agree it would be very good to use Mersenne Twister for the pseudo-random numbers. 

If you are happy to look at my PR do you want that behind a feature flag or do you prefer Mersenne Twister replaces rand by default?

**Comment by petr-tik** [comment]
I have been looking at your library, while learning about derivatives pricing and I have a question. 

In your options pricing model you use the widespread Black-Scholes model with a normal distribution. I have read and heard about different flavours of Black-Scholes using other probability distributions eg. fat-tailed.  

https://github.com/MarcusRainbow/QuantMath/blob/master/src/math/optionpricing.rs

Is there any scope/use for adding/trying other pdf for this?

**Comment by MarcusRainbow** [comment]
Hi,

Please could you point me to your pull request and I'll take a look.
Personally, I'd be happy for Mersenne Twister to replace rand by default,
though it would be useful to also support other pseudo-random and
low-discrepancy sequences.

*  acknowledging the moat of C++ and integrating rust with existing
codebases.*

Yes, that is a problem, though as you can see I have Rust roughly
integrating with C++ in the library. My preference is to have a facade
layer around an analytics library anyway. If you were working in a bank,
there would generally be one team doing the analytics and an IT team doing
the surrounding databases, real-time feeds and connections to other IT
systems (Goldman Sachs is not like this, but their IT is pretty broken
anyway). It is therefore very useful to have a clear division between the
analytics and the surrounding IT -- ideally one that allows serialisation
of the objects passed between the layers, so you can easily debug the
analytics or the IT without having to debug one system inside the other.

Thanks,
Marcus

**Comment by MarcusRainbow** [comment]
In terms of using other pdfs for option pricing, yes that can be useful. In
practice, options are "priced" using vanilla black scholes, which is not
really a model at all.

It would be a model if the volatility used was consistent across times and
strikes, but by making it strike and time dependent, effectively we are
just saying the "volatility" is just an expression of the price.
Calculating the price is really just unpacking it from the so-called vol
surface. That is what the black scholes formula is used for. At best, it is
a way of interpolating prices between strikes and expiry dates, though even
then you have to be really careful with dividends and vol events.

For this reason, BNP Paribas use a local vol model for pricing even vanilla
options. However, it is a pain in the neck for them.

All that said, the more formulae the better. They all have their uses.

Marcus

**Comment by petr-tik** [comment]
I started poking at it - haven't got a PR yet, but have experienced problems building with the current version of rustc as well as failing tests. 

Judging from previous commits you must run CI locally, so I guess flaky tests are ok?

**Comment by MarcusRainbow** [comment]
Yes I've had trouble with tests not reproducing on different architectures.
That was the tests involving pseudorandom numbers. Hopefully changing to
Mersenne Twister will address that. Are the other tests ok for you?
Thanks,
Marcus
